File: 5.binarySearch/441.ArrangingCoins/main.go
package main

// https://leetcode.com/problems/arranging-coins/

// sum = n * (n + 1) / 2

func arrangeCoins(n int) int {
	low := 1
	high := n

	for low <= high {
		mid := low + (high-low)/2
		sum := mid * (mid + 1) / 2
		if sum == n {
			return mid
		}
		if sum > n {
			high = mid - 1
		} else {
			low = mid + 1
		}
	}

	return high
}

func arrangeCoins2(n int) int {
	low := 1
	high := n

	for low <= high {
		mid := low + (high-low)/2
		sum := mid * (mid + 1) / 2
		if sum == n {
			return mid
		}
		if sum > n {
			high = mid - 1
		} else {
			low = mid + 1
		}
	}

	return low - 1
}

File: 5.binarySearch/441.ArrangingCoins/main.py
# https://leetcode.com/problems/arranging-coins/

# 1 - 1
# 2 - 3
# 3 - 6
# 4 - 10
# => 1 + 2 + 3 + 4 = 10
# sum = n * (n + 1) / 2
# TC: O(logn), SC: O(1)
class Solution:
    def arrangeCoins(self, n: int) -> int:
        l, r = 1, n
        res = 0
        while l <= r:
            m = l + (r-l)//2
            s = m * (m + 1) // 2
            if n == s:
                return m
            if s < n:
                l = m + 1
                res = m
            else:
                r = m - 1
        return res
